### PR TITLE
Diagnostic event handlers

### PR DESCRIPTION
**Pull request by elhigu**
This is work in progress PR for initial comments, with only documentation and implementation for some events.

Missing: 

* create / destroy events
* tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -97,20 +97,51 @@ try {
...
 
 // how many asynchronous create calls are running
-pool.numPendingCreates()
+pool.numPendingCreates();
 
 // waits for all resources to be returned to the pool and destroys them.
 // pool cannot be used after this.
 await pool.destroy();
+
+// Adds synchronous event handlers to for example to allow externally
```
Comment (koskimas) [inline README.md]: Better wording:

The following examples add synchronous event handlers for example to allow ...

**Comment by bttmly** [comment]
I think you could avoid subtle but surprising differences from "real" Node.js event emitters by embedding an emitter in the pool object (e.g. `this._emitter = new EventEmitter()`) and forwarding `on` calls from the pool to the underlying emitter. Then using `this._emitter.emit()`. You'd also be able to get stuff like removing handlers for "free" by forwarding those calls as well.

**Comment by elhigu** [comment]
@bttmly seems like a good idea. Though I probably still  need to wrap those event handlers to some function that catches and logs errors, so providing remove for each single handler might not be so straight forward.

**Comment by bttmly** [comment]
I think if you just always call some internal method rather than `emitter.emit` manually, you're ok?

```js
class Pool {
constructor () {
  this._emitter = new EventEmitter();
  // rest of init logic
}

on (...args) {
  this._emitter.on(...args);
}

removeListener (...args) {
  this._emitter.removeListener(...args);
}

// only this method is allowed to call this._emitter.emit() !
_executeEventHandlers (...args) {
  try {
    this._emitter.emit(...args);
  } catch (err) {
    // do warning
  }
}
}
```

**Comment by elhigu** [comment]
@bttmly that is not really exactly the same thing... wouldn't in that case single invalid event handler throwing an error stop execution of all the rest of registered handlers?

**Comment by bttmly** [comment]
I would think that's fine, that is the expected behavior of Node event emitters. Since we're deferring a lot to the caller already (e.g. timers, etc) I don't think providing an extra layer of error safety, beyond what's necessary to keep the library's current behavior, is necessary.

**Comment by elhigu** [comment]
@bttmly I'll implement it in a way that I can get both: fail safety and ability to remove listeners. So I wont be calling emit directly, but just get list of listeners from _emitter and call each listener separately, like it is done right now.

**Comment by elhigu** [comment]
Need to fix compatibility to node 8+ level at least... knex 0.18 will not have node 6 support anymore.

**Comment by elhigu** [comment]
I think this is fine now, but probably should not be released before knex 0.18 is out. Also more review comments are welcome.

**Comment by bttmly** [comment]
This looks awesome

**Pull request merged**